drivers/dma/dma_cavs_gpdma.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT intel_cavs_gpdma
+
+#define GPDMA_CTL_OFFSET 0x004
+#define GPDMA_CTL_FDCGB BIT(0)
+
+#include "dma_dw_common.h"
+
+#define LOG_LEVEL CONFIG_DMA_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(dma_cavs_gpdma);
+
+
+/* Device run time data */
+struct cavs_gpdma_data {
+	struct dw_dma_dev_data dw_data;
+};
+
+/* Device constant configuration parameters */
+struct cavs_gpdma_cfg {
+	struct dw_dma_dev_cfg dw_cfg;
+	uint32_t shim;
+};
+
+/* Disables automatic clock gating (force disable clock gate) */
+static void cavs_gpdma_clock_enable(const struct device *dev)
+{
+	const struct cavs_gpdma_cfg *const dev_cfg = dev->config;
+	uint32_t reg = dev_cfg->shim + GPDMA_CTL_OFFSET;
+
+	sys_write32(GPDMA_CTL_FDCGB, reg);
+}
+
+
+int cavs_gpdma_init(const struct device *dev)
+{
+	const struct cavs_gpdma_cfg *const dev_cfg = dev->config;
+
+	/* Disable dynamic clock gating appropriately before initializing */
+	cavs_gpdma_clock_enable(dev);
+
+	/* Disable all channels and Channel interrupts */
+	dw_dma_setup(dev);
+
+	/* Configure interrupts */
+	dev_cfg->dw_cfg.irq_config();
+
+	LOG_INF("Device %s initialized", dev->name);
+
+	return 0;
+}
+
+
+static const struct dma_driver_api cavs_gpdma_driver_api = {
+	.config = dw_dma_config,
+	.reload = dw_dma_reload,
+	.start = dw_dma_transfer_start,
+	.stop = dw_dma_transfer_stop,
+};
+
+#define CAVS_GPDMA_INIT(inst)						\
```

Comment:
This is a BIG macro.  Maybe it might be cleaner to do several FOREACH iterations over the devicetree instances, e.g. one to define the config structs, one to declare interrupts, etc...?

Reply:
That's a fair statement, mostly people write the one macro to init all the things for each instance, but this is quite big. Perhaps splitting it up at least might make it easier to read through, and perhaps share more with the original dw_dma.c